### PR TITLE
feat: indicate client with valid certificate [WPB-3220]

### PR DESCRIPTION
**Pull request by theswiftway**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-3220" title="WPB-3220" target="_blank">StoryWPB-3220</a>  [iOS] 5.3.14 Indicate client with valid certificate
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Solutions

Changes the layout for displaying Proteus and MLS information on Client list screen when navigating from the Conversation screen. Also displays Proteus verification and E2E Identity status images

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test
Steps
1. Go to any conversation
2. Tap on other user
3. Tap on Devices Tab
4. Relevant IDs and images should be shown based on if it is a proteus protocol or MLS protocol.

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)
IMG_0006

IMG_0007

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios/Wire-iOS/Sources/Helpers/UserClient+Fingerprint.swift [hunk 1]
```diff
@@ -52,7 +52,7 @@ extension UserClientType {
         let identifierPrefixString = L10n.Localizable.Registration.Devices.id + " "
         let identifierString = NSMutableAttributedString(string: identifierPrefixString, attributes: attributes)
         let identifier = uppercase ? displayIdentifier.localizedUppercase : displayIdentifier
-        let attributedRemoteIdentifier = identifier.fingerprintStringWithSpaces.fingerprintString(attributes: attributes, boldAttributes: boldAttributes)
+        let attributedRemoteIdentifier = identifier.fingerprintStringWithSpaces.fingerprintString(attributes: attributes, boldAttributes: [:])
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/Helpers/UserClient+Fingerprint.swift]: question: Why do we remove setting the boldAttributes?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift [hunk 2]
```diff
@@ -23,142 +23,174 @@ import WireDataModel
...
-    private let deviceTypeIconView = UIImageView()
+    // MARK: - Properties
+    let nameLabel = DynamicFontLabel(style: .headline,
+                                     color: LabelColors.textDefault)
+    let labelLabel = DynamicFontLabel(style: .subheadline,
+                                      color: LabelColors.textDefault)
+    let mlsThumbprintLabel = DynamicFontLabel(style: .caption1,
+                                        color: LabelColors.textCellSubtitle)
+    let proteusIdLabel = DynamicFontLabel(style: .caption1,
+                                            color: LabelColors.textCellSubtitle)
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: 👍 Thank you for getting rid of old initializers and using the new ones! On that note: Did you check how it would look like on a larger font size and if we need to make any changes there?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift [hunk 3]
```diff
@@ -23,142 +23,174 @@ import WireDataModel
...
-    private let subtitleLabel = DynamicFontLabel(fontSpec: .mediumRegularFont,
-                                                 color: LabelColors.textCellSubtitle)
+    typealias LabelColors = SemanticColors.Label
+    typealias IconColors = SemanticColors.Icon
 
-    private let deviceTypeIconView = UIImageView()
+    // MARK: - Properties
+    let nameLabel = DynamicFontLabel(style: .headline,
+                                     color: LabelColors.textDefault)
+    let labelLabel = DynamicFontLabel(style: .subheadline,
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: suggestion: `labelLabel` isn't descriptive enough. I would suggest a name like: `userClientLabel`

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift [hunk 4]
```diff
@@ -23,142 +23,174 @@ import WireDataModel
...
-
-    private weak var client: UserClientType?
-
-    override func setUp() {
-        super.setUp()
+    var showLabel: Bool = false {
+        didSet {
+            updateLabel()
+        }
+    }
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: suggestion: showLabel isn't descriptive enough: I would suggest something like: `showUserClientLabel`

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift [hunk 5]
```diff
@@ -23,142 +23,174 @@ import WireDataModel
...
-        subtitleLabel.attributedText = client.attributedRemoteIdentifier(attributes,
-                                                                         boldAttributes: boldAttributes,
-                                                                         uppercase: true)
+    private func updateLabel() {
+        if let userClientLabel = viewModel?.label, showLabel {
+            labelLabel.text = userClientLabel
+        } else {
+            labelLabel.text = ""
+        }
+    }
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: suggestion: `updateLabel` isn't descriptive enough: I would suggest something like: `updateUserClientLabel`

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift [hunk 6]
```diff
@@ -23,142 +23,174 @@ import WireDataModel
...
+    required init(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
 
+    // MARK: - Methods
+    func setupStyle() {
+        nameLabel.accessibilityIdentifier = "device name"
+        labelLabel.accessibilityIdentifier = "device label"
+        proteusIdLabel.accessibilityIdentifier = "device proteus ID"
+        mlsThumbprintLabel.accessibilityIdentifier = "device mls thumbprint"
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: **suggestion**: Changing or adding any accessibilityIdentifiers will impact and may disrupt QA during their own testing. Can you please inform QA that those changes are to happen so they can update their code too?

**Review by agisilaos — CHANGES_REQUESTED** [review]
@theswiftway I left some questions/suggestions. Can we please have some snapshots tests for those screens? If we already have snapshot tests can we make sure that they're up to date?

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift (hunk 6 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: Noted

File: wire-ios/Wire-iOS/Sources/Helpers/UserClient+Fingerprint.swift (hunk 1 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/Helpers/UserClient+Fingerprint.swift]: reverted this

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift (hunk 2 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: Yup. Looks fine

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift (hunk 4 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift (hunk 5 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift (hunk 3 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: done

Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/UserProfile/Devices/UserClientCell.swift]: don't need it anymore

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png:1]: The position of the shield seems to be wrong here.

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png:1]: The position of the shield seems to be wrong here.

File: wire-ios/Wire-iOS Tests/UserClientCellTests.swift [hunk 7]
```diff
@@ -20,64 +20,125 @@ import XCTest
...
-        configuration(cell)
-        cell.layoutIfNeeded()
-        return cell
+    override func setUp() {
+        super.setUp()
+        let frame = CGRect(x: 0, y: 0, width: 390, height: 84)
+        sut = UserClientCell(frame: frame)
+        container = containerView(with: sut, snapshotBackgroundColor: nil)
+        container.frame = frame
     }
```
Comment (agisilaos) [inline wire-ios/Wire-iOS Tests/UserClientCellTests.swift:33]: Since we do have a setUp method it would be a good idea to have a tearDown as well and set our properties to nil.

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift [hunk 8]
```diff
@@ -22,25 +22,18 @@ import Contacts
 import WireDataModel
 import WireCommonComponents
 
-final class ClientTableViewCell: UITableViewCell, DynamicTypeCapable {
```
Comment (agisilaos) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift:25]: If you would like to remove DynamicTypeCapable from this class please make sure all of the labels here use the new initializer from DynamicFontLabel. 

`let nameLabel = DynamicFontLabel(fontSpec: .normalSemiboldFont, color: LabelColors.textDefault)` still uses the old initializer

**Review by agisilaos — CHANGES_REQUESTED** [review]
Left a few more comments. @theswiftway I just realized the color of the ! in this screenshot is wrong. Can you please update the color to match the design specs?

Screenshot 2024-02-16 at 15 26 24

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png:1]: https://www.figma.com/file/gSmh7BRyq9SGKkyOLSbOdM/E2E-Identity?type=design&node-id=456-35017&mode=design&t=BIRFAddKVDsMzNtn-0

It seems fine

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift (hunk 8 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift:25]: All the labels use `DynamicFontLabel`

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png:1]: Its just the way snapshot test takes the snapshot. I have attached images in the PR. Please refer that.

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift (hunk 8 above)
Comment (agisilaos) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift:25]: I agree but this specific label uses the deprecated initializer. If you would like to remove conformance to DynamicTypeCapable you need to make sure you're using the new initializers on all labels.

The new one would look like this:

`let nameLabel = DynamicFontLabel(style: .body , color: LabelColors.textDefault)` // .body is just an example, please make sure we're using the correct font style.

File: wire-ios/Wire-iOS Tests/UserClientCellTests.swift (hunk 7 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/UserClientCellTests.swift:33]: done

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png:1]: In the mock it seems to be centered to the text. On this snapshot it does not. Did we miss anything with the constraints?

Screenshot 2024-02-16 at 15 44 47

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png
Comment (agisilaos) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png:1]: In the mock it seems to be centered to the text. On this snapshot it does not. Did we miss anything with the constraints?

Screenshot 2024-02-16 at 15 47 51

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift (hunk 8 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift:25]: got it. Done

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png:1]: Its just a stackview snapshot issue. It seems to misalign. If you check the attached images in the PR. There is an example which shows right placement.

File: wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png
Comment (theswiftway) [reply wire-ios/Wire-iOS Tests/ReferenceImages/UserClientCellTests/testThatProteusBadgeIsDisplayed_whenProteusIsVerified_inDarkMode.1.png:1]: It seems to be more to be an issue with snapshot library for testing `UIStackView`

**Comment by theswiftway** [comment]
> * testThatSharedInstanceReturnQualityTypeWifi

Not sure why this is getting triggered 🤔

**Review by johnxnguyen — APPROVED** [review]

File: wire-ios-data-model/Source/Model/UserClient/UserClientType.swift [hunk 9]
```diff
@@ -39,6 +39,9 @@ import Foundation
 
     var needsToNotifyUser: Bool { get set }
 
+    var mlsThumbPrint: String? { get set }
+
+    var e2eIdentityCertificate: E2eIdentityCertificate? { get set }
```
Comment (KaterinaWire) [inline wire-ios-data-model/Source/Model/UserClient/UserClientType.swift:44]:
```suggestion
    var e2eIdentityCertificate: E2eIdentityCertificate? { get set }

```

File: wire-ios/Tests/Mocks/MockConversation.swift [hunk 10]
```diff
@@ -74,6 +74,8 @@ class SwiftMockConversation: NSObject, Conversation {
     var areServicesPresent: Bool = false
 
     var domain: String?
+
+    var isConversationEligibleForVideoCalls = false
```
Comment (KaterinaWire) [inline wire-ios/Tests/Mocks/MockConversation.swift]: Do we use this property? Looks like it's not related to this PR

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift [hunk 11]
```diff
@@ -520,6 +513,7 @@ final class ClientListViewController: UIViewController,
                     for client in userClients {
                         let mlsClientIdRawValue = mlsClients[client.clientId.hashValue]?.rawValue
                         client.e2eIdentityCertificate = certificates.first { $0.clientId == mlsClientIdRawValue }
+                        client.mlsThumbPrint = client.e2eIdentityCertificate?.mlsThumbprint ?? client.mlsPublicKeys.ed25519
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift]: **suggestion**: I see you are repeating `client.e2eIdentityCertificate?.mlsThumbprint ?? client.mlsPublicKeys.ed25519` several times. Let's reduce the repetitive code and create a helper property.

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift [hunk 12]
```diff
@@ -159,18 +132,21 @@ extension ClientTableViewCellModel {
 
     private typealias DeviceDetailsSection = L10n.Localizable.Device.Details.Section
 
-    static func from(userClient: UserClient) -> ClientTableViewCellModel {
-        let title = userClient.isLegalHoldDevice ?  L10n.Localizable.Device.Class.legalhold : (userClient.model ?? "")
-
+    static func from(userClient: UserClientType, shouldSetType: Bool = true) -> ClientTableViewCellModel {
+        var title = ""
+        if shouldSetType {
+            title = userClient.deviceClass == .legalHold ?  L10n.Localizable.Device.Class.legalhold : (userClient.deviceClass?.localizedDescription.capitalized ?? userClient.type.localizedDescription.capitalized)
```
Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift]: It's not clear from the code why not to use `userClient.type.localizedDescription.capitalized` every time. Why do we need to check if `userClient.deviceClass` is nil. I would create a helper method or property

Comment (KaterinaWire) [inline wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift]:
```suggestion
            title = userClient.deviceClass == .legalHold ? L10n.Localizable.Device.Class.legalhold : (userClient.deviceClass?.localizedDescription.capitalized ?? userClient.type.localizedDescription.capitalized)
```

**Comment by KaterinaWire** [comment]
@arthurwolf There are broken snapshots `testThatProteusBadgeIsDisplayed_whenProteusIsVerified.1.png` and in the dark mode.

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift (hunk 12 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift]: This was the way it was done. I am merely applying the same logic as before

File: wire-ios-data-model/Source/Model/UserClient/UserClientType.swift (hunk 9 above)
Comment (theswiftway) [reply wire-ios-data-model/Source/Model/UserClient/UserClientType.swift:44]: I don't see whats the change here @KaterinaWire

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift (hunk 11 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientListViewController.swift]: done

File: wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift (hunk 12 above)
Comment (theswiftway) [reply wire-ios/Wire-iOS/Sources/UserInterface/Settings/DeviceManagement/ClientTableViewCell.swift]: done

File: wire-ios/Tests/Mocks/MockConversation.swift (hunk 10 above)
Comment (theswiftway) [reply wire-ios/Tests/Mocks/MockConversation.swift]: Good catch. Removed it

**Review by agisilaos — APPROVED** [review]
